crates/holochain_state/src/source_chain.rs

```diff
@@ -300,10 +283,77 @@ impl SourceChain {
         let author = self.author.clone();
         let persisted_head = self.head_info.as_ref().map(|h| h.action.clone());
 
+        let now = Timestamp::now();
+
+        // Take out a write lock as late as possible, after doing everything we can in memory and
+        // before starting any database read/write operations.
+        let write_permit = self.vault.acquire_write_permit().await?;
+
+        self.vault
+            .read_async({
+                let author = author.clone();
+                move |txn| {
+                    let chain_lock = get_chain_lock(&txn, author.as_ref())?;
+                    match chain_lock {
+                        Some(chain_lock) => {
+                            // If the chain is locked, the lock must be for this entry.
+                            if chain_lock.subject() != lock_subject {
+                                return Err(SourceChainError::ChainLocked);
+                            }
+                            // If the lock is expired then we can't write this countersigning session.
+                            else if chain_lock.is_expired_at(now) {
+                                return Err(SourceChainError::LockExpired);
+                            }
+
+                            // Otherwise, the lock matches this entry and has not expired. We can proceed!
+                        }
+                        None => {
+                            // If this is a countersigning entry but there is no chain lock then maybe
+                            // the session expired before the entry could be written or maybe the app
+                            // has just made a mistake. Either way, it's not valid to write this entry!
+                            if is_countersigning_session {
+                                return Err(SourceChainError::CountersigningWriteWithoutSession);
+                            }
+                        }
+                    }
+
+                    Ok(())
+                }
+            })
+            .await?;
+
+        // Sync with CHC, if CHC is present
+        if let Some(chc) = network.chc() {
+            // Skip the CHC sync if this is a countersigning session.
+            // If the session times out, we might roll the chain back to the previous head, and so
+            // we don't want the record to exist remotely.
+            if is_countersigning_session {
+                tracing::debug!(
+                    "Skipping CHC push for countersigning session: {:?}",
+                    records
+                );
+            } else {
+                let payload = AddRecordPayload::from_records(
+                    self.keystore.clone(),
+                    (*self.author).clone(),
+                    records,
+                )
+                .await
+                .map_err(SourceChainError::other)?;
+
+                match chc.add_records_request(payload).await {
```

Comment:
Just to check my understanding:

1. If this call errors, the flush will fail.
2. Other Holochain tasks cannot write to the database while this call is happening, even if it is slow.

Reply:
Yes and yes. Which for example, means that receiving validation receipts during a a slow CHC flush will then also  be slow because it will have to wait for the  write permit to be released. It may impact the throughput of the system a little on writes.